lib/versioning/pep440/index.spec.ts

```diff
@@ -87,7 +87,7 @@ describe('versioning/pep440/index', () => {
     currentValue          | rangeStrategy | currentVersion | newVersion | expected
     ${'1.0.0'}            | ${'bump'}     | ${'1.0.0'}     | ${'1.2.3'} | ${'1.2.3'}
     ${'1.0.0'}            | ${'replace'}  | ${'1.0.0'}     | ${'1.2.3'} | ${'1.2.3'}
-    ${'1.0.0'}            | ${'pin'}      | ${'1.0.0'}     | ${'1.2.3'} | ${'==1.2.3'}
+    ${'1.0.0'}            | ${'pin'}      | ${'1.0.0'}     | ${'1.2.3'} | ${'1.2.3'}
```

Comment:
This looks wrong to me. as pin means to add `==` for pep440. You maybe need to disable `pin` to fix #12348 ?

Reply:
you maybe sinmply need to configure `rangeStrategy=replace`, as the test above shows that it would do what you want.

Reply:
Yes, why is this necessary if `replace` already does what you need?

Reply:
Hmmm I did read through all the Versioning and Managers documentation but I can't see anything in there that says `pep440` versioning with `rangeStrategy=pin` must prepend an operator.  Is that recorded elsewhere in the docs that I missed?  If not, can we get the docs updated to mention that `pep440` versioning with `rangeStrategy=pin` will always prepend the operator, and clarify the usage of `rangeStrategy=replace` with `pep440` versioning for cases where operator prepending is not desired?  I do agree that most cases for `pep440` would have the operator prepended but it is definitely not wanted for `pre-commit` manager.

How would this work for `regexManagers:dockerfileVersions` if I want to use `rangeStrategy=pin`?  I think I should change my Dockerfile to match this format?  This will move the operator from inside `PIP_DEPS` to be inside each `X_VERSION` ARG.
```Dockerfile
# renovate: datasource=pypi depName=nox versioning=pep440
ARG NOX_VERSION="==2021.6.12"
# renovate: datasource=pypi depName=dunamai versioning=pep440
ARG DUNAMAI_VERSION="==1.6.0"
# renovate: datasource=pypi depName=packaging versioning=pep440
ARG PACKAGING_VERSION="==21.0"
# renovate: datasource=pypi depName=wheel versioning=pep440
ARG WHEEL_VERSION="==0.37.0"
# renovate: datasource=pypi depName=twine versioning=pep440
ARG TWINE_VERSION="==3.4.2"
# renovate: datasource=pypi depName=cython versioning=pep440
ARG CYTHON_VERSION="==0.29.24"
# renovate: datasource=pypi depName=numpy versioning=pep440
ARG NUMPY_VERSION="==1.21.2"
# renovate: datasource=pypi depName=scipy versioning=pep440
ARG SCIPY_VERSION="==1.7.1"
ARG PIP_DEPS="nox${NOX_VERSION} dunamai${DUNAMAI_VERSION} packaging${PACKAGING_VERSION} wheel${WHEEL_VERSION} twine${TWINE_VERSION} cython${CYTHON_VERSION} numpy${NUMPY_VERSION} scipy${SCIPY_VERSION}"
```
If so, it would help to have the documentation improved on this aspect too.  I was originally using the example at https://docs.renovatebot.com/modules/manager/regex/#advanced-capture before the `regexManagers:dockerfileVersions` preset was added and I now realise I was obviously confused by the lack of documentation regarding `pep440` versioning always prepending the operator.

Also, would it be possible to update the `regexManagers:dockerfileVersions` preset and associated advanced capture example to include a `rangeStrategy` in the special comment?  Not every dependency inside a Dockerfile will have the same versioning and be coming from the same source so they may well need a different `rangeStrategy` too.

Reply:
Why do you insist on using rangeStrategy=pin? It's meant for managers which support ranges, and regex manager does not.

Reply:
Hmm ok.  When we first started using renovate, we read https://docs.renovatebot.com/dependency-pinning/ and decided "yes, pinning looks like what we want to use everywhere", even though the article refers to nodejs the ideas seem applicable to other language package management ecosystems too.  In addition to that, we read through the default presets and saw [:pinVersions](https://docs.renovatebot.com/presets-default/#pinversions) which looked like what we wanted.  We also saw [:preserveSemverRanges](https://docs.renovatebot.com/presets-default/#preservesemverranges) and after reading that it mentioned "preserving semver ranges", we thought that `rangeStrategy=replace` wasn't applicable to pinning versions.  The docs for [rangeStrategy=replace](https://docs.renovatebot.com/configuration-options/#rangestrategy) also mentioned "replacing ranges" which seemed to be not related to the "pinning" that we had decided to use.  In addition, we never saw anything in the documentation to say that `rangeStrategy=pin` is for supporting ranges and that ranges are not compatible with `regexManager`.  Everything we have read in the docs is suggesting that `rangeStrategy=pin` is for pinning and `rangeStrategy=replace` is for ranges.

Based on what you've said we should be using `:preserveSemverRanges` for our regexManager `pep440` version pinning instead of `:pinVersions`?

Reply:
Good point! Yes, this is a scenario where you need to change the strategy specifically for this manager

Reply:
Thanks!

Reply:
So there is nothing that needs changing in renovate, except perhaps some docs clarification? If not, then I guess this PR and issue can be closed?

Reply:
Also, is it possible to add a `rangeStrategy` template into the `regexManagers:dockerfileVersions` preset regex?  I don't see it listed on https://docs.renovatebot.com/modules/manager/regex/ so if not, can it be changed through a `packageRules` section like this?
```json
"packageRules": [
  {
    "extends": "regexManagers:dockerfileVersions",
    "rangeStrategy": "replace"
  }
]
```
I would like to leave `:pinVersions` at the top of our renovate config.json if possible and only set `rangeStrategy=replace` where strictly necessary.